keps/sig-network/20200406-graceful-termination-external-traffic-policy-local.md

```diff
@@ -0,0 +1,190 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Services with `externalTrafficPolicy=Local` lack the ability to gracefully handle traffic from a loadbalancer when it goes from N to 0 endpoints.
+Since terminating pods are never considered "ready" in Endpoints/EndpointSlice, a node with only terminating endpoints would drop traffic even though
+it may still be part of a loadbalancer's node pool. Even with loadbalancer health checks, there is usually a delay between when the health check
+fails and when a node is completely decommissioned.
+
+This KEP proposes that if all endpoints for a given Service (with externalTrafficPolicy=Local) in any given node are `Terminating`, then traffic for this Service
+should be redirected to those terminating endpoints. This ensures that traffic is not dropped between the time a node fails its health check (has 0 endpoints) and
+when the node is decommissioned from the loadbalancer's node pool.
+
+## Motivation
+
+### Goals
+
+* enable zero downtime rolling updates for Services with ExternalTrafficPolicy=Local via node ports or loadbalancer IPs.
+
+### Non-Goals
+
+* changing the behavior of terminating pods/endpoints outside the scope of Services with ExternalTrafficPolicy=Local via a nodeport/loadbalancerIPs/externalIPs.
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user I would like to do a rolling update of a Deployment fronted by a Service Type=LoadBalancer with ExternalTrafficPolicy=Local.
+If a node that has only 1 pod of said deployment goes into the `Terminating` state, all traffic to that node is dropped until either a new pod
+comes up or my cloud provider removes the node from the loadbalancer's node pool. Ideally the terminating pod should gracefully handle traffic to this node
+until one of the above conditions are satifised.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Additions to EndpointSlice
+
+Today, terminating endpoints are treated the same as "not ready" endpoints. In order to make a distinction between an endpoint that is terminating
+and an endpoint that has failed readiness checks, a `Terminating` field would be added to `EndpointCondition` as part of the EndpointSlice API.
+
+No additions will be made to Endpoints meaning this change will depend on the EndpointSlice API being enabled.
```

Comment:
Can I make this assumption or should we also add terminating metadata to Endpoints even though kube-proxy will likely be using EndpointSlice pretty soon?